prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,369 @@
+
+	return err;
+
+out_err:
+	if (base_ep->efa_recv_wr_index > 0) {
+		post_recv_err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+		if (OFI_UNLIKELY(post_recv_err)) {
+			EFA_WARN(FI_LOG_EP_DATA,
+				 "Encountered error %ld when ibv_post_recv on error handling path\n",
+				 post_recv_err);
+		}
+	}
+
+	base_ep->efa_recv_wr_index = 0;
+
+	return err;
+}
+
+static ssize_t efa_ep_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+
+	return efa_post_recv(base_ep, msg, flags);
+}
+
+static ssize_t efa_ep_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+			   void *desc, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct iovec iov;
+	struct fi_msg msg;
+
+	EFA_SETUP_IOV(iov, buf, len);
+	EFA_SETUP_MSG(msg, &iov, &desc, 1, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static ssize_t efa_ep_recvv(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
+			    size_t count, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct fi_msg msg;
+
+	EFA_SETUP_MSG(msg, iov, desc, count, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static inline ssize_t efa_post_send(struct efa_base_ep *base_ep, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_qp *qp = base_ep->qp;
+	struct efa_conn *conn;
+	struct ibv_sge sg_list[2];  /* efa device support up to 2 iov */
+	struct ibv_data_buf inline_data_list[2];
+	size_t len, i;
+	int ret = 0;
+
+	dump_msg(msg, "send");
+
+	conn = efa_av_addr_to_conn(base_ep->av, msg->addr);
+	assert(conn && conn->ep_addr);
+
+	assert(msg->iov_count <= base_ep->info->tx_attr->iov_limit);
+
+	len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+
+	if (qp->ibv_qp->qp_type == IBV_QPT_UD) {
+		assert(msg->msg_iov[0].iov_len >= base_ep->info->ep_attr->msg_prefix_size);
+		len -= base_ep->info->ep_attr->msg_prefix_size;
+	}
+
+	assert(len <= base_ep->info->ep_attr->max_msg_size);
+
+	if (!base_ep->is_wr_started) {
+		ibv_wr_start(qp->ibv_qp_ex);
+		base_ep->is_wr_started = true;
+	}
+
+	qp->ibv_qp_ex->wr_id = (uintptr_t)msg->context;
+	if (flags & FI_REMOTE_CQ_DATA) {
+		ibv_wr_send_imm(qp->ibv_qp_ex, msg->data);
+	} else {
+		ibv_wr_send(qp->ibv_qp_ex);
+	}
+
+	if (len <= base_ep->domain->device->efa_attr.inline_buf_size &&
+	    (!msg->desc || !efa_mr_is_hmem(msg->desc[0]))) {
+		for (i = 0; i < msg->iov_count; i++) {
+			inline_data_list[i].addr = msg->msg_iov[i].iov_base;
+			inline_data_list[i].length = msg->msg_iov[i].iov_len;
+
+			/* Whole prefix must be on the first sgl for dgram */
+			if (!i && qp->ibv_qp->qp_type == IBV_QPT_UD) {
+				inline_data_list[i].addr = (char*)inline_data_list[i].addr + base_ep->info->ep_attr->msg_prefix_size;
```

Comment:
I remember we used to  skip the first 1 iov if its length is the same with the prefix size...But IIRC having a 0 length sg_list[0] should be fine as driver will skip them